src/toast/todmap/filterbin.py

```diff
@@ -0,0 +1,725 @@
+                if self._zip_maps:
+                    fname += ".gz"
+                white_noise_cov.write_healpix_fits(fname)
+                if self.rank == 0:
+                    print(
+                        "OpFilterBin: Wrote white noise covariance to {}"
+                        "".format(fname),
+                        flush=True,
+                    )
+
+        dist_map.allreduce()
+        covariance_apply(white_noise_cov, dist_map)
+
+        fname = os.path.join(self._outdir, self._outprefix + suffix + ".fits")
+        if self._zip_maps:
+            fname += ".gz"
+        dist_map.write_healpix_fits(fname)
+        if self.rank == 0:
+            print("OpFilterBin: Wrote map to {}".format(fname), flush=True)
+
+        return white_noise_cov
+
+    @function_timer
+    def _get_detweights(self, data):
+        # FIXME : We need a generic way of building the detector weights
+        # FIXME : This can wait until TOAST3.
+        detweights = {}
+        for obs in data.obs:
+            tod = obs["tod"]
+            for det in tod.local_dets:
+                if det not in detweights:
+                    detweights[det] = 1e3
+        return detweights
+
+    def _load_deprojection_map(self, data):
+        if self._deproject_map is None:
+            return None
+        deprojection_map = DistPixels(
+            data,
+            comm=self.comm,
+            nnz=self._deproject_nnz,
+            dtype=np.float32,
+            pixels=self._pixels_name,
+        )
+        deprojection_map.read_healpix_fits(self._deproject_map)
+        return deprojection_map
+
+    @function_timer
+    def exec(self, data, comm=None):
+
+        if comm is None:
+            self.comm = data.comm.comm_world
+        else:
+            self.comm = comm
+        if self.comm is None:
+            self.rank = 0
+            self.ntask = 1
+        else:
+            self.rank = self.comm.rank
+            self.ntask = self.comm.size
+
+        # Filter data
+
+        obs_matrix = self._initialize_obs_matrix()
+        detweights = self._get_detweights(data)
+        deprojection_map = self._load_deprojection_map(data)
+
+        white_noise_cov = None
+        if self._write_binned:
+            white_noise_cov = self._bin_map(data, detweights, "binned")
+
+        t0 = time()
+        t1 = time()
+        if self.verbose > 1 and self.rank == 0:
+            print("OpFilterBin: Filtering signal", flush=True)
+
+        for obs in data.obs:
+            tod = obs["tod"]
+            if self._intervals in obs:
+                intervals = obs[self._intervals]
+            else:
+                intervals = None
+            local_intervals = tod.local_intervals(intervals)
+            times = tod.local_times()
+            common_flags = tod.local_common_flags(self._common_flag_name)
+
+            phase = self._get_phase(tod)
+            t1 = time()
+            common_templates = self._build_common_templates(
```

Comment:
More notes for toast3 porting...  It seems like the `FilterBin` operator could be constructed like:

1. operator which builds common templates for every observation and view
2. operator which builds and regresses per-detector templates
3. operator which accumulates observation matrix
4. Pipeline which runs pointing expansion and (2.) - (3.) one detector at a time

Then the `FilterBin` operator can internally run (1.) and then `BinMap` passing the pipeline in (4.) as the `pre_process` option.  Anyway, I'm sure we'll iterate on this along the way.